routing/payment_lifecycle.go

```diff
@@ -0,0 +1,339 @@
+	htlcAdd := &lnwire.UpdateAddHTLC{
+		Amount:      route.TotalAmount,
+		Expiry:      route.TotalTimeLock,
+		PaymentHash: p.payment.PaymentHash,
+	}
+	copy(htlcAdd.OnionBlob[:], onionBlob)
+
+	// Attempt to send this payment through the network to complete
+	// the payment. If this attempt fails, then we'll continue on
+	// to the next available route.
+	firstHop := lnwire.NewShortChanIDFromInt(
+		route.Hops[0].ChannelID,
+	)
+
+	// We generate a new, unique payment ID that we will use for
+	// this HTLC.
+	paymentID, err := p.router.cfg.NextPaymentID()
+	if err != nil {
+		return err
+	}
+
+	// We now have all the information needed to populate
+	// the current attempt information.
+	p.attempt = &channeldb.PaymentAttemptInfo{
+		PaymentID:  paymentID,
+		SessionKey: sessionKey,
+		Route:      *route,
+	}
+
+	// Before sending this HTLC to the switch, we checkpoint the
+	// fresh paymentID and route to the DB. This lets us know on
+	// startup the ID of the payment that we attempted to send,
+	// such that we can query the Switch for its whereabouts. The
+	// route is needed to handle the result when it eventually
+	// comes back.
+	err = p.router.cfg.Control.RegisterAttempt(p.payment.PaymentHash, p.attempt)
+	if err != nil {
+		return err
+	}
+
+	log.Tracef("Attempting to send payment %x (pid=%v), "+
+		"using route: %v", p.payment.PaymentHash, paymentID,
+		newLogClosure(func() string {
+			return spew.Sdump(route)
+		}),
+	)
+
+	// Send it to the Switch. When this method returns we assume
+	// the Switch successfully has persisted the payment attempt,
+	// such that we can resume waiting for the result after a
+	// restart.
+	err = p.router.cfg.Payer.SendHTLC(
+		firstHop, paymentID, htlcAdd,
+	)
+	if err != nil {
+		log.Errorf("Failed sending attempt %d for payment "+
+			"%x to switch: %v", paymentID, p.payment.PaymentHash, err)
+		return err
+	}
+
+	log.Debugf("Payment %x (pid=%v) successfully sent to switch",
+		p.payment.PaymentHash, paymentID)
+
+	return nil
+}
+
+// handleSendError inspects the given error and determines whether we should
+// make another payment attempt.
+func (p *paymentLifecycle) handleSendError(sendErr error) error {
+	// Failed before we created a payment attempt is final.
+	if p.attempt == nil {
+		return sendErr
+	}
+
+	finalOutcome := p.router.processSendError(
+		p.paySession, &p.attempt.Route, sendErr,
+	)
+
+	if finalOutcome {
+		log.Errorf("Payment %x failed with final outcome: %v",
+			p.payment.PaymentHash, sendErr)
+
+		// Mark the payment failed with no route.
+		// TODO(halseth): make payment codes for the actual reason we
+		// don't continue path finding.
+		err := p.router.cfg.Control.Fail(
+			p.payment.PaymentHash, channeldb.FailureReasonNoRoute,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Terminal state, return the error we encountered.
+		return sendErr
+	}
+
+	// We get ready to make another payment attempt.
+	p.lastError = sendErr
+	p.attempt = nil
```

Comment:
It seems to me that checking and nil-ing the attempt shouldn't happen in this function, but in the main loop.

Reply:
I agree. Also noticed that the current logic was a bit wrong, since we would call `control.Fail` more than once for some errors.

Fixed by moving the `sendAttempt` into its own method, such that `handleSendError` is only called for errors returned by the switch: https://github.com/lightningnetwork/lnd/pull/2761/commits/1de31e3a3afbb148f9132f3c960a9bfbc9236594